internal/logger/logger_test.go

```diff
@@ -39,6 +45,114 @@ func BenchmarkLogger(b *testing.B) {
 	})
 }
 
+// mockKeyValues will return a slice of alternating keys and values of the
+// given length and the resulting JSON string.
+func mockKeyValuesB(b *testing.B, length int) KeyValues {
+	b.Helper()
+
+	keysAndValues := KeyValues{}
+	for i := 0; i < length; i++ {
+		keyName := fmt.Sprintf("key%d", i)
+		valueName := fmt.Sprintf("value%d", i)
+
+		keysAndValues.Add(keyName, valueName)
+	}
+
+	return keysAndValues
+}
+
+func BenchmarkIOSinkInfo(b *testing.B) {
+	keysAndValues := mockKeyValuesB(b, 100)
+
+	b.ReportAllocs()
+	b.ResetTimer()
+
+	sink := NewIOSink(bytes.NewBuffer(nil))
+
+	for i := 0; i < b.N; i++ {
+		sink.Info(0, "foo", keysAndValues...)
+	}
+}
+
+func mockKeyValuesT(t *testing.T, length int) (KeyValues, map[string]interface{}) {
+	t.Helper()
+
+	keysAndValues := KeyValues{}
+	m := map[string]interface{}{}
+
+	for i := 0; i < length; i++ {
+		keyName := fmt.Sprintf("key%d", i)
+		valueName := fmt.Sprintf("value%d", i)
+
+		keysAndValues.Add(keyName, valueName)
+		m[keyName] = valueName
+	}
+
+	return keysAndValues, m
+}
+
+type mockIOWriter struct {
+	msgs chan []byte
+}
```

Comment:
Optional: A more real-world use case would be to use something like a `bytes.Buffer` as a writer, then try to read the stream of JSON messages from the byte buffer using a `json.Decoder`. The `bytes.Buffer` would play the role of the file or output stream (e.g. STDOUT), which is what the `IOSink` will be writing to under real-world conditions. Also, that removes the need to implement a custom `io.Writer`.

See an example in the Go Playground [here](https://go.dev/play/p/68aiAvdZ_C3).